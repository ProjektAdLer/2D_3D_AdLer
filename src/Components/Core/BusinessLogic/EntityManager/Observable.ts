export default abstract class Observable<T> {
  protected value: T;
  protected subscribers: ((value: T) => void)[] = [];

  public subscribe(fn: (value: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (value: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  protected notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  get Value(): T {
    return this.value;
  }
}
