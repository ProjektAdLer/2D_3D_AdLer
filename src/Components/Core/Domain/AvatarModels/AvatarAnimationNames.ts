const AvatarAnimationNames = {
  // animations for avatar
  idle: "ac_anim_idle2_avatarSkeleton",
  walk: "ac_anim_walkcycle_avatarSkeleton",
  interact: "ac_anim_interact_avatarSkeleton",
  // animations for npcs
  npc_idle: "ac_anim_idle2",
  npc_walk: "ac_anim_walkcycle",
} as const;

export default AvatarAnimationNames;
