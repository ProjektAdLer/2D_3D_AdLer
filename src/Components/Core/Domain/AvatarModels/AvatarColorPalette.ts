/**
 * Represents a color in the avatar color palette.
 */
export interface AvatarColor {
  /**
   * The id of the color on the color palette.
   */
  id: number;

  /**
   * The key for the name of the color, used for localization.
   */
  nameKey: string;

  /**
   * The hexadecimal representation of the color.
   */
  hexColor: string;
}

const AvatarColorPalette: AvatarColor[] = [
  // Black
  { id: 0, nameKey: "Black 1", hexColor: "#000000" },
  { id: 1, nameKey: "Black 2", hexColor: "#1b1b1b" },
  { id: 2, nameKey: "Black 3", hexColor: "#313131" },
  { id: 3, nameKey: "Black 4", hexColor: "#555555" },
  { id: 4, nameKey: "Black 5", hexColor: "#767676" },
  { id: 5, nameKey: "Black 6", hexColor: "#949494" },
  { id: 6, nameKey: "Black 7", hexColor: "#b3b3b3" },
  { id: 7, nameKey: "Black 8", hexColor: "#cecece" },
  // Blue
  { id: 8, nameKey: "Blue 1", hexColor: "#043675" },
  { id: 9, nameKey: "Blue 2", hexColor: "#043675" }, // duplicate
  { id: 10, nameKey: "Blue 3", hexColor: "#2f91cd" },
  { id: 11, nameKey: "Blue 4", hexColor: "#83aece" },
  { id: 12, nameKey: "Blue 5", hexColor: "#17607f" },
  { id: 13, nameKey: "Blue 6", hexColor: "#5398b0" },
  { id: 14, nameKey: "Blue 7", hexColor: "#94b7c3" },
  { id: 15, nameKey: "Blue 8", hexColor: "#b4c6ce" },
  // Red
  { id: 16, nameKey: "Red 1", hexColor: "#8b1518" },
  { id: 17, nameKey: "Red 2", hexColor: "#a41e22" },
  { id: 18, nameKey: "Red 3", hexColor: "#bd2c30" },
  { id: 19, nameKey: "Red 4", hexColor: "#ce4145" },
  { id: 20, nameKey: "Red 5", hexColor: "#871a0b" },
  { id: 21, nameKey: "Red 6", hexColor: "#a8411e" },
  { id: 22, nameKey: "Red 7", hexColor: "#c1643f" },
  { id: 23, nameKey: "Red 8", hexColor: "#cdada2" },
  // Orange
  { id: 24, nameKey: "Orange 1", hexColor: "#b76000" },
  { id: 25, nameKey: "Orange 2", hexColor: "#c98300" },
  { id: 26, nameKey: "Orange 3", hexColor: "#cea100" },
  { id: 27, nameKey: "Orange 4", hexColor: "#ceb753" },
  { id: 28, nameKey: "Orange 5", hexColor: "#aa9501" },
  { id: 29, nameKey: "Orange 6", hexColor: "#c2af09" },
  { id: 30, nameKey: "Orange 7", hexColor: "#ccc428" },
  { id: 31, nameKey: "Orange 8", hexColor: "#cccaa5" },
  // Brown
  { id: 32, nameKey: "Brown 1", hexColor: "#25160c" },
  { id: 33, nameKey: "Brown 2", hexColor: "#4b2a1a" },
  { id: 34, nameKey: "Brown 3", hexColor: "#633d29" },
  { id: 35, nameKey: "Brown 4", hexColor: "#835b32" },
  { id: 36, nameKey: "Brown 5", hexColor: "#bb9871" },
  { id: 37, nameKey: "Brown 6", hexColor: "#a6947d" },
  { id: 38, nameKey: "Brown 7", hexColor: "#b1a28b" },
  { id: 39, nameKey: "Brown 8", hexColor: "#bba989" },
  // more Brown
  { id: 40, nameKey: "Brown 9", hexColor: "#7d4337" },
  { id: 41, nameKey: "Brown 10", hexColor: "#b97947" },
  { id: 42, nameKey: "Brown 11", hexColor: "#ac7d59" },
  { id: 43, nameKey: "Brown 12", hexColor: "#677f88" },
  { id: 44, nameKey: "Brown 13", hexColor: "#b29676" },
  { id: 45, nameKey: "Brown 14", hexColor: "#bfa183" },
  { id: 46, nameKey: "Brown 15", hexColor: "#c9b2a6" },
  { id: 47, nameKey: "Brown 16", hexColor: "#ccbeb8" },
  // violet
  { id: 48, nameKey: "Violet 1", hexColor: "#a909b3" },
  { id: 49, nameKey: "Violet 2", hexColor: "#c55eca" },
  { id: 50, nameKey: "Violet 3", hexColor: "#c8a6cb" },
  { id: 51, nameKey: "Violet 4", hexColor: "#cbc0cc" },
  { id: 52, nameKey: "Violet 5", hexColor: "#500c8c" },
  { id: 53, nameKey: "Violet 6", hexColor: "#863fb3" },
  { id: 54, nameKey: "Violet 7", hexColor: "#a370c3" },
  { id: 55, nameKey: "Violet 8", hexColor: "#bdaacb" },
  // green
  { id: 56, nameKey: "Green 1", hexColor: "#0d5502" },
  { id: 57, nameKey: "Green 2", hexColor: "#199a09" },
  { id: 58, nameKey: "Green 3", hexColor: "#3ac61e" },
  { id: 59, nameKey: "Green 4", hexColor: "#82ce84" },
  { id: 60, nameKey: "Green 5", hexColor: "#3b6613" },
  { id: 61, nameKey: "Green 6", hexColor: "#709636" },
  { id: 62, nameKey: "Green 7", hexColor: "#9db16b" },
  { id: 63, nameKey: "Green 8", hexColor: "#bdc6a5" },
];

export default AvatarColorPalette;
