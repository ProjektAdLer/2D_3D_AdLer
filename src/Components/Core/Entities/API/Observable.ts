export default class Observable<T> {
  constructor(data?: T) {
    if (data) {
      this.data = data;
    }
  }
  private data: T;
  private subscribers: Array<(data: T) => void> = [];

  public subscribe(fn: (data: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (data: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  public notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  public getData(): T {
    return this.data;
  }

  public setData(data: T): void {
    this.data = data;
    this.notify(data);
  }
}
