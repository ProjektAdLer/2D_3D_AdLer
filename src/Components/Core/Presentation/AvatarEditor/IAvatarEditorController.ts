import AvatarConfigTO from "../../Application/DataTransferObjects/AvatarConfigTO";

export default interface IAvatarEditorController {
  onAvatarConfigChanged(changes: Partial<AvatarConfigTO>): void;
  saveAvatarConfig(): void;
  resetAvatarConfig(): void;
}
