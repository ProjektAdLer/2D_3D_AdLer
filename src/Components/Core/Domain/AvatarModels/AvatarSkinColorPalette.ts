import { AvatarColor } from "./AvatarColorPalette";

const AvatarSkinColorPalette: AvatarColor[] = [
  { id: 0, nameKey: "Dark 1", hexColor: "#4f2a1a", uOffset: 0, vOffset: 0 },
  { id: 1, nameKey: "Dark 2", hexColor: "#6e3b2f", uOffset: 0.125, vOffset: 0 },
  { id: 2, nameKey: "Dark 3", hexColor: "#8a5733", uOffset: 0.25, vOffset: 0 },
  { id: 3, nameKey: "Dark 4", hexColor: "#a36f4f", uOffset: 0.375, vOffset: 0 },
  { id: 4, nameKey: "Light 1", hexColor: "#c08e6c", uOffset: 0.5, vOffset: 0 },
  {
    id: 5,
    nameKey: "Light 2",
    hexColor: "#d8a980",
    uOffset: 0.625,
    vOffset: 0,
  },
  { id: 6, nameKey: "Light 3", hexColor: "#eac9a4", uOffset: 0.75, vOffset: 0 },
  {
    id: 7,
    nameKey: "Light 4",
    hexColor: "#f3dcc6",
    uOffset: 0.875,
    vOffset: 0,
  },
];
export default AvatarSkinColorPalette;
