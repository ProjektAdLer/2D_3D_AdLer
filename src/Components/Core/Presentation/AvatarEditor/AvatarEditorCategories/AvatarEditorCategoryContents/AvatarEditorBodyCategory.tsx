import { useState } from "react";
import AvatarEditorCategoryContentProps from "./AvatarEditorCategoryContentProps";
import { useTranslation } from "react-i18next";
import ColorPickerButton from "~ReactComponents/GeneralComponents/ColorPicker/ColorPickerButton";
import RangeSlider from "~ReactComponents/ReactRelated/ReactBaseComponents/RangeSlider";
import ColorPickerModal from "~ReactComponents/GeneralComponents/ColorPicker/ColorPickerModal";
import AvatarColorPalette from "src/Components/Core/Domain/AvatarModels/AvatarColorPalette";
// TODO: update Icons
import bodySmallIcon from "../../../../../../Assets/icons/body-small.svg";
import bodyBigIcon from "../../../../../../Assets/icons/body-big.svg";

export default function AvatarEditorBodyCategory(
  props: AvatarEditorCategoryContentProps,
) {
  const { t: translate } = useTranslation("avatarEditor");
  const [showModal, setShowModal] = useState(false);
  const [skinColor, setSkinColor] = useState(AvatarColorPalette[0]);

  return (
    <div className="flex flex-col">
      <div className="pb-2 border-b border-gray-500">
        <h1 className="text-2xl font-bold">{translate("bodyColorTitle")}</h1>
      </div>
      <div className="w-full p-2 m-2">
        <ColorPickerButton
          currentColor={skinColor}
          onClick={() => setShowModal(true)}
        />
        <ColorPickerModal
          title={translate("bodyColorTitle")}
          showModal={showModal}
          onClose={() => setShowModal(false)}
          onColorClickFunction={(color) => {
            setSkinColor(color);
          }}
        />
      </div>
      <div className="pb-2 border-b border-gray-500">
        <h1 className="text-2xl font-bold">
          {translate("bodyProportionsTitle")}
        </h1>
      </div>

      <div className="bodyProportionContainer">
        <div className="flex items-center">
          <h3 className="text-xl font-bold">{translate("upperBodyTitle")}</h3>
        </div>
        <div>
          <RangeSlider
            min={0}
            max={1}
            step={0.01}
            callback={console.log}
            displayFactor={100}
            buttons={{ imageLeft: bodySmallIcon, imageRight: bodyBigIcon }}
          />
        </div>

        <div className="flex items-center">
          <h3 className="text-xl font-bold">{translate("lowerBodyTitle")}</h3>
        </div>
        <div>
          <RangeSlider
            min={0}
            max={1}
            step={0.01}
            callback={console.log}
            displayFactor={100}
            buttons={{ imageLeft: bodySmallIcon, imageRight: bodyBigIcon }}
          />
        </div>
      </div>
    </div>
  );
}
