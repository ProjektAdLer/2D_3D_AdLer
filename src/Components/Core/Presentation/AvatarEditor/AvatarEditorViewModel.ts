import Observable from "src/Lib/Observable";
import { AvatarColor } from "../../Domain/AvatarModels/AvatarColorPalette";
import {
  AvatarHairModels,
  AvatarBeardModels,
  AvatarHeadgearModels,
  AvatarGlassesModels,
  AvatarShirtModels,
  AvatarPantsModels,
  AvatarShoesModels,
  AvatarBackpackModels,
  AvatarOtherModels,
} from "../../Domain/AvatarModels/AvatarModelTypes";

export default class AvatarEditorViewModel {
  // Face
  eyebrows: Observable<number> = new Observable<number>(0);
  eyes: Observable<number> = new Observable<number>(0);
  nose: Observable<number> = new Observable<number>(0);
  mouth: Observable<number> = new Observable<number>(0);

  // Hair
  hair: Observable<AvatarHairModels> = new Observable<AvatarHairModels>("none");
  beard: Observable<AvatarBeardModels> = new Observable<AvatarBeardModels>(
    "none",
  );
  hairColor: Observable<AvatarColor> = new Observable<AvatarColor>({
    id: 33,
    nameKey: "Brown 2",
    hexColor: "#4b2a1a",
  });

  // Accessories
  headgear: Observable<AvatarHeadgearModels> =
    new Observable<AvatarHeadgearModels>("none");
  glasses: Observable<AvatarGlassesModels> =
    new Observable<AvatarGlassesModels>("none");
  backpack: Observable<AvatarBackpackModels> =
    new Observable<AvatarBackpackModels>("none");
  other: Observable<AvatarOtherModels> = new Observable<AvatarOtherModels>(
    "none",
  );

  // Clothes
  shirt: Observable<AvatarShirtModels>;
  shirtColor: Observable<AvatarColor>;
  pants: Observable<AvatarPantsModels>;
  pantsColor: Observable<AvatarColor>;
  shoes: Observable<AvatarShoesModels>;
  shoesColor: Observable<AvatarColor>;

  // Body
  skinColor: Observable<AvatarColor> = new Observable<AvatarColor>({
    id: 33,
    nameKey: "Brown 2",
    hexColor: "#4b2a1a",
  });
  roundness: Observable<number>; // 0-1 morph target weight

  // --------------------------------------------------------------
  // UI Component Visibility
  // --------------------------------------------------------------

  uiVisiblity: AvatarEditorUI = {
    hairMenu: {
      hairstyles: new Observable<boolean>(true),
      beards: new Observable<boolean>(true),
    },
    faceMenu: {
      eyebrows: new Observable<boolean>(true),
      eyes: new Observable<boolean>(false),
      noses: new Observable<boolean>(false),
      mouths: new Observable<boolean>(false),
    },
    accessoireMenu: {
      headGear: new Observable<boolean>(true),
      glasses: new Observable<boolean>(false),
      backpack: new Observable<boolean>(false),
      other: new Observable<boolean>(false),
    },
    clothingMenu: {
      shirts: new Observable<boolean>(true),
      pants: new Observable<boolean>(false),
      shoes: new Observable<boolean>(false),
    },
  };
}

export interface AvatarConfigHairIconVisibility {
  hairstyles: Observable<boolean>;
  beards: Observable<boolean>;
}

export interface AvatarConfigFaceIconVisibility {
  eyebrows: Observable<boolean>;
  eyes: Observable<boolean>;
  noses: Observable<boolean>;
  mouths: Observable<boolean>;
}

export interface AvatarConfigAccessoireIconVisibility {
  headGear: Observable<boolean>;
  glasses: Observable<boolean>;
  backpack: Observable<boolean>;
  other: Observable<boolean>;
}

export interface AvatarConfigClothingIconVisibility {
  shirts: Observable<boolean>;
  pants: Observable<boolean>;
  shoes: Observable<boolean>;
}

export interface AvatarEditorUI {
  hairMenu: AvatarConfigHairIconVisibility;
  faceMenu: AvatarConfigFaceIconVisibility;
  accessoireMenu: AvatarConfigAccessoireIconVisibility;
  clothingMenu: AvatarConfigClothingIconVisibility;
}
