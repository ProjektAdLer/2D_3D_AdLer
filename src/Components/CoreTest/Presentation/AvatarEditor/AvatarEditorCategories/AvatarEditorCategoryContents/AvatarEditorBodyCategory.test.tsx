import { mock } from "jest-mock-extended";
import AvatarEditorBodyCategory from "../../../../../Core/Presentation/AvatarEditor/AvatarEditorCategories/AvatarEditorCategoryContents/AvatarEditorBodyCategory";
import AvatarEditorViewModel from "../../../../../Core/Presentation/AvatarEditor/AvatarEditorViewModel";
import IAvatarEditorController from "../../../../../Core/Presentation/AvatarEditor/IAvatarEditorController";
import Observable from "../../../../../../Lib/Observable";
import { fireEvent, render } from "@testing-library/react";
import { AvatarColor } from "../../../../../Core/Domain/AvatarModels/AvatarColorPalette";
import React from "react";
import AvatarSkinColorPalette from "../../../../../Core/Domain/AvatarModels/AvatarSkinColorPalette";

const avatarEditorMock = mock<AvatarEditorViewModel>();
const avatarEditorControllerMock = mock<IAvatarEditorController>();

describe("AvatarEditorBodyCategory", () => {
  test("should render", () => {
    avatarEditorMock.skinColor = new Observable<AvatarColor>({
      id: 33,
      nameKey: "Brown 2",
      hexColor: "#4b2a1a",
    });

    const { container } = render(
      <AvatarEditorBodyCategory
        viewModel={avatarEditorMock}
        controller={avatarEditorControllerMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  test("click on colorpicker-button opens color picker", () => {
    const container = render(
      <AvatarEditorBodyCategory
        viewModel={avatarEditorMock}
        controller={avatarEditorControllerMock}
      />,
    );

    const colorpickerButton = container.getByText("Brown 2");
    fireEvent.click(colorpickerButton);

    expect(container.getAllByText("bodyColorTitle")[0]).toBeInTheDocument();
  });

  test("click on color in colorpicker calls controller", () => {
    avatarEditorMock.skinColor = new Observable<AvatarColor>(
      AvatarSkinColorPalette[4],
    );
    const container = render(
      <AvatarEditorBodyCategory
        viewModel={avatarEditorMock}
        controller={avatarEditorControllerMock}
      />,
    );

    const colorpickerButton = container.getByText("Light 1");
    fireEvent.click(colorpickerButton);

    const color = container.getByTestId("Light 1");
    fireEvent.click(color);

    expect(avatarEditorControllerMock.onAvatarConfigChanged).toHaveBeenCalled();
  });
});
