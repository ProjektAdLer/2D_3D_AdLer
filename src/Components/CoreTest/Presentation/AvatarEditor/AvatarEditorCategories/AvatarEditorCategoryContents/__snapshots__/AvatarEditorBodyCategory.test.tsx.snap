// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarEditorBodyCategory should render 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="pb-2 border-b border-gray-500"
    >
      <h1
        class="text-2xl font-bold"
      >
        bodyColorTitle
      </h1>
    </div>
    <div
      class="w-full p-2 m-2"
    >
      <div
        class=""
      >
        <div
          class=""
        >
          <button
            class="px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
            data-testid="colorpicker"
            title="ColorButtonToolTip"
          >
            <div
              class="border-2 border-black rounded m-2 h-5 w-5"
              style="background-color: rgb(75, 42, 26);"
            />
            Brown 2
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
